sdks/python/src/opik/integrations/dspy/callback.py

```diff
@@ -0,0 +1,99 @@
+from typing import Any, Dict, Optional
+
+from dspy.utils.callback import BaseCallback
+
+from opik import context_storage, opik_context
+from opik.api_objects import opik_client, span, trace
+from opik.decorator import error_info_collector
+
+
+class OpikCallback(BaseCallback):
+
+    def __init__(
+        self,
+        project_name: Optional[str] = None,
+    ):
+        self._opik_trace_data: Optional[trace.TraceData] = None
+        self._project_name = project_name
+        self._opik_client = opik_client.Opik(
+            _use_batching=True,
+            project_name=project_name,
+        )
+
+    def on_module_start(
+        self,
+        call_id: str,
+        instance: Any,
+        inputs: Dict[str, Any],
+    ):
+        print(f"on_module_start() is called with call_id: {call_id}, instance: {instance.__class__.__name__}, inputs: {inputs}")
+
+        if self._opik_trace_data is not None:
+            return
+
+        existing_trace_data = opik_context.get_current_trace_data()
+        if existing_trace_data:
+            self._opik_trace_data = existing_trace_data
+        else:
+            trace_data = trace.TraceData(
+                name=instance.__class__.__name__,
+                metadata={"created_from": "dspy"},
+                project_name=self._project_name,
+            )
+            self._opik_trace_data = trace_data
+
+    def on_module_end(
+        self,
+        call_id: str,
+        outputs: Optional[Any],
+        exception: Optional[Exception] = None,
+    ):
+        print(f"on_module_end() is called with call_id: {call_id}, outputs: {outputs}, exception: {exception}")
+
+        if self._opik_trace_data is None:
+            return
+
+        if exception:
+            error_info = error_info_collector.collect(exception)
+            self._opik_trace_data.update(error_info=error_info)
+
+        self._opik_trace_data.init_end_time()
+        self._opik_client.trace(**self._opik_trace_data.__dict__)
+        self._opik_trace_data = None
+
+    def on_lm_start(
+        self,
+        call_id: str,
+        instance: Any,
+        inputs: Dict[str, Any],
+    ):
+        print(f"LM is called with inputs: {inputs}")
+
+        span_data = span.SpanData(
+            trace_id=self._opik_trace_data.id,
+            name=instance.__class__.__name__,
+            parent_span_id=None,
+            type="llm",
+            input=inputs,
+            project_name=self._opik_trace_data.project_name,
+        )
+        context_storage.add_span_data(span_data)
```

Comment:
Working with the context storage like that is unsafe, we risk totally break the library logic if for example we create a span and then due to some exception it is left in the context.

The way it should work is: all spans and traces created inside this callback should be stored inside this callback. The only interaction with the opik_context is allowed when we create the first span/trace and want to check of there is already an existing one to attach to.

`context_storage` is a very low-level and sensitive to unhappy flows part of the library, it should almost never be manipulated like that manually unless it is localized and we 100% control the execution flow.